fuzz/fuzz_types.ml

```diff
@@ -0,0 +1,440 @@
+    unwrap field_type wrapped
+
+(** Dynamically build an Irmin variant (assuming it has 1 to 5 cases). *)
+and irmin_variant : string -> case list -> dyn_variant T.t = [%impl_variant 5]
+
+(** Dynamically build an Irmin enum (assuming it has 1 to 5 cases). *)
+and irmin_enum : string -> string list -> dyn_variant T.t =
+ fun n ts -> irmin_variant n (List.map (fun t -> (t, ACT Case0)) ts)
+
+(** Create a generator for values of a given dynamic type. *)
+let rec t_to_value_gen : type a. a t -> a gen = function
+  | TUnit -> const ()
+  | TBool -> bool
+  | TChar -> char
+  | TInt -> int
+  | TInt32 -> int32
+  | TInt64 -> int64
+  | TFloat -> float
+  | TString -> string
+  | TBytes -> bytes
+  | TList ty -> list (t_to_value_gen ty)
+  | TArray ty -> map [ list (t_to_value_gen ty) ] Array.of_list
+  | TOption ty -> option (t_to_value_gen ty)
+  | TPair (ty1, ty2) -> pair (t_to_value_gen ty1) (t_to_value_gen ty2)
+  | TTriple (ty1, ty2, ty3) ->
+      triple (t_to_value_gen ty1) (t_to_value_gen ty2) (t_to_value_gen ty3)
+  | TResult (ty1, ty2) -> result (t_to_value_gen ty1) (t_to_value_gen ty2)
+  | TRecord (n, fs) -> irmin_record_gen n fs
+  | TVariant (n, cs) -> irmin_variant_gen n cs
+  | TEnum (n, ts) -> irmin_enum_gen n ts
+
+(** Create a generator for an Irmin record with the given specification. *)
+and irmin_record_gen : string -> field list -> dyn_record gen =
+ fun record_name fields ->
+  let rec fields_gen = function
+    | [] -> const []
+    | (field_name, AT field_type) :: xs ->
+        map [ t_to_value_gen field_type; fields_gen xs ] (fun v fs' ->
+            (field_name, wrap field_type v) :: fs')
+  in
+  map [ fields_gen fields ] (fun fs' -> new_dyn_record record_name fs')
+
+(** Create a generator for an Irmin variant with the given specification. *)
+and irmin_variant_gen : string -> case list -> dyn_variant gen =
+ fun variant_name cases ->
+  let rec cases_gen = function
+    | [] -> []
+    | (case_name, ACT Case0) :: xs ->
+        const (variant_name, case_name, VUnit ()) :: cases_gen xs
+    | (case_name, ACT (Case1 t)) :: xs ->
+        map [ t_to_value_gen t ] (fun v -> (variant_name, case_name, wrap t v))
+        :: cases_gen xs
+  in
+  choose (cases_gen cases)
+
+(** Create a generator for an Irmin enum with the given specification. *)
+and irmin_enum_gen : string -> string list -> dyn_variant gen =
+ fun n ts -> irmin_variant_gen n (List.map (fun t -> (t, ACT Case0)) ts)
+
+type inhabited = Inhabited : 'a t * 'a -> inhabited
+
+(** Generate a [Inhabited (t, v)], where [t : 'a t] is a dynamic type and
+    [v : 'a] is a corresponding value. *)
+let inhabited_gen : inhabited gen =
+  dynamic_bind any_type_gen @@ fun (AT t) ->
+  map [ t_to_value_gen t ] (fun v -> Inhabited (t, v))
+
+(** Pretty-print a value of a given dynamic type. [~context] can be used to
+    print additional context in case of an error. *)
+let pp_value :
+    type a. ?context:(unit -> string) -> a t -> Format.formatter -> a -> unit =
```

Comment:
I'm wondering, is there a reason to use `type a.` instead of `'a`?

Reply:
The former is stronger, enforcing that the function is truly polymorphic in `a` (i.e. it has the same meaning as an `'a` in the signature). When `'a` is used in a constraint, it simply means 'any type' and permits the constrained function to be monomorphic:

```ocaml
let increment_int : int -> int = fun a -> a

let increment : 'a -> 'a = increment_int
(* SUCCESS, because: unify(int, 'a) holds when 'a = int *)

let increment : type a. a -> a = increment_int
(* FAIL, because: unify(∀ a. a -> a, int -> int) has no solution `)
```'

Reply:
Thanks for the explanation! 
The unification occurs when the function is applied, right? Here, `pp_value` is applied on values of the same type (in `bin_check` and `json_check`) and indeed it compiles with the "weaker" type signature. Is there a convention on which one to use, when both work?

Reply:
In this case, the code will compile with any of the following weaker constraints:

- `?context:(unit -> string) -> 'a t -> Format.formatter -> 'a -> unit`
- `?context:'a -> 'b -> 'c -> 'd -> unit`
- `?context:'a -> 'b`

since they all unify to the most-general type of `∀ a. ?context:(unit -> string) -> a t -> Format.formatter -> a -> unit`. Generally, adding a constraint doesn't prevent the type-checker from inferring a more general type: it's just a lower bound on the generality of the type. In this case, the usages in `bin_check` and `json_check` both _require_ this function to be truly polymorphic, since they apply it to a value taken from `Inhabited` (which may have any type), so it makes sense for the constraint to be the strictest one possible.

Generally,

 - The `type a.` form is for saying "this value must be polymorphic" in a constraint. For instance, the constraint in `let ignore_pair : type a b. (a, b) -> unit` says "this function must ignore pairs of any component types". (This is what `'a` means when used in a signature, which is confusing.)

 - The `'a` form is for saying "I don't care" in a constraint. For instance, the constraint in `let x : ('a, int) = ("foo", 0)` says "Value `x` must be a pair with `int` as its second component, and I don't care about the first". The type-checker still infers a type `(string, int)` in this case, and will use that in type-checking occurrences of the value `x` elsewhere.

Because of this, the `type a.` form is generally more useful in structures.

Aside: in the presence of GADTs the type-checker is often not able to infer polymorphic types, so it's necessary to use constraints to guide it to finding the most-general type (c.f. [usages in `Irmin.Type`](https://github.com/mirage/irmin/blob/master/src/irmin/type/irmin_type.ml#L150)). In this context, the `'a` form of constraint won't save you since it's not _requiring_ polymorphism (the type-checker doesn't get the guidance it needs).

Reply:
It makes sense now, thanks a lot!